### ISSUE TITLE
attn_fwd persistent loop skips every tile after a fully-masked early exit (continue_condition never reset)

### ISSUE THREAD

**Issue by jeffdaily** [ROCm/aotriton#235]
## Summary

In `tritonsrc/fwd_kernel.py` (0.13b; the same code is on `main` at `modules/flash/kernel/fwd_kernel.py`, lines 236/271/344), `continue_condition` is initialised to `True` once, before the persistent `while tile_id < num_tiles_total` loop, and the causal early exit for fully-masked tiles (`if (lb_empty and fb_empty) and rb_empty: ... continue_condition = False`) never resets it. With `PERSISTENT_TYPE == 2` (dynamic, the mode PyTorch selects for causal masks by passing `persistent_atomic_counter`), a workgroup that draws a fully-masked tile and then acquires another tile from the atomic counter skips that tile entirely: no `Out` rows and no `L` (logsumexp) entries are written, so those rows keep whatever the output buffers held.

Whether a workgroup acquires a second tile depends on scheduling: with a few tiles and hundreds of workgroups each normally gets one tile, but when another process shares the GPU a workgroup that finished a trivial fully-masked tile can win the next `atomic_add` before a fresh workgroup starts. That makes the corruption appear only under GPU contention.

## Observed

gfx950 (MI355X), ROCm 10.0 (HIP 7.15), PyTorch main, AOTriton 0.13b as bundled by PyTorch, fp16, bottom-right causal (`CAUSAL_TYPE == 2`), `seqlen_q=192, seqlen_k=64` (rows 0..127 fully masked), batch 1, 2 heads, head_dim 64 / 128 / 160 / 192 / 256.

- Alone on the GPU: 0 of 25 runs corrupted; all 256 masked-row logsumexp entries are `inf` and the masked output rows are 0, as designed.
- With a second process running 6144x6144 GEMMs on the same GPU: 2 to 3 of 25 runs corrupted per head_dim (23 of 25 in a tighter loop through `torch.nn.functional.scaled_dot_product_attention` with `causal_lower_right`), each with exactly one row of NaN in `Out` at the causal boundary (row 129, 130, 132) or row 0, and only 32 to 128 of the 256 masked-row logsumexp entries equal to `inf` (the rest untouched).
- `AMD_SERIALIZE_KERNEL=3` does not change the rate (kernel-internal, not cross-kernel).
- The CK backend (`torch.backends.cuda.preferred_rocm_fa_library("ck")`) is clean under the same contention.
- Passing a null `persistent_atomic_counter` from PyTorch to force non-persistent scheduling was not a usable bisection (the kernel dereferences the counter); the source analysis below is the evidence, plus the symptom match: the number of masked-row logsumexp entries that are `inf` drops from 256 to 32..128 exactly as skipped fully-masked tiles would predict.

## Deterministic reproducer (no co-tenant needed)

The same defect shows deterministically as soon as there are more tiles than workgroups, because then every workgroup loops over several tiles and any that starts with a fully-masked tile skips the rest. Same shape per head (`seqlen_q=192, seqlen_k=64`, head_dim 64, fp16, bottom-right causal), larger batch/heads, single process, nothing else on the GPU (gfx950):

| batch x heads | tiles (3 per head) | corrupted runs | masked-row logsumexp entries equal to `inf` (expected) |
|---|---|---|---|
| 1 x 2 | 6 | 0 / 10 | 256 (256) |
| 8 x 16 | 384 | 10 / 10 | 10944 (16384) |
| 32 x 16 | 1536 | 10 / 10 | 15552..19968 (65536), and in some runs 32000 masked output rows left non-zero |
| 64 x 32 | 6144 | 10 / 10 | 16320..16384 (262144), and in some runs 245760 masked output rows left non-zero |

The roughly constant ~16k written entries at the two largest sizes is the number of tiles the workgroups process before the flag is poisoned. Script:

```python
import torch, sys
B, H, q_len, k_len, head_dim = int(sys.argv[1]), int(sys.argv[2]), 192, 64, 64
n = int(sys.argv[3])
bad = 0; details = []
for i in range(n):
    torch.manual_seed(i)
    q = torch.randn(B, H, q_len, head_dim, device="cuda", dtype=torch.float16)
    k = torch.randn(B, H, k_len, head_dim, device="cuda", dtype=torch.float16); v = torch.randn_like(k)
    out, lse, *_ = torch.ops.aten._efficient_attention_forward(q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2), None, None, None, None, None, 0.0, 2, True)
    out = out.transpose(1, 2)
    masked = out[:, :, :q_len - k_len]
    expect_inf = B * H * (q_len - k_len)
    inf = torch.isinf(lse).sum().item(); nan = torch.isnan(out).any(-1).sum().item(); nz = (masked != 0).any(-1).sum().item()
    if inf != expect_inf or nan or nz:
        bad += 1; details.append(f"iter {i}: lse inf {inf}/{expect_inf}, NaN rows {nan}, nonzero masked rows {nz}")
print(f"B={B} H={H} tiles={B*H*3}: corrupted runs {bad}/{n}" + ("" if not details else "\n  " + "\n  ".join(details[:4])), flush=True)
```

Run as `python masked_rows_manytiles.py 32 16 10` (batch, heads, runs).

## Reproducer under contention (the original symptom)

Two scripts on one GPU: start `gpu_cotenant.py` in the background, then run `python masked_rows_diag.py 160 40` (PyTorch main, `HIP_VISIBLE_DEVICES` pinned to the same GPU). Alone, `masked_rows_diag.py` reports `none` every time.

`masked_rows_diag.py`:

```python
import torch, sys
from torch.nn.attention.bias import causal_lower_right
q_len, k_len, head_dim, n = 192, 64, int(sys.argv[1]), int(sys.argv[2])
lib = sys.argv[3] if len(sys.argv) > 3 else None
if lib: torch.backends.cuda.preferred_rocm_fa_library(lib)
bias = causal_lower_right(q_len, k_len)
seen = {}
for i in range(n):
    torch.manual_seed(42)
    q = torch.randn(1, 2, q_len, head_dim, device="cuda", dtype=torch.float16)
    k = torch.randn(1, 2, k_len, head_dim, device="cuda", dtype=torch.float16); v = torch.randn_like(k)
    # low-level op with the bottom-right causal mask type (what sdpa does for CausalBias LOWER_RIGHT)
    out, lse, seed, offset, mq, mk = torch.ops.aten._efficient_attention_forward(
        q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2), None, None, None, None, None, 0.0, 2, True)  # custom_mask_type 2 = CausalFromBottomRight
    out = out.transpose(1, 2)
    nan_rows = torch.isnan(out).any(-1)  # [1, 2, q_len]
    if nan_rows.any():
        rows = sorted(set(nan_rows[0].nonzero()[:, 1].tolist()))
        key = (len(rows), rows[0], rows[-1], torch.isnan(lse).sum().item(), torch.isinf(lse).sum().item())
        seen[key] = seen.get(key, 0) + 1
masked = q_len - k_len
print(f"head_dim={head_dim} backend={torch.backends.cuda.preferred_rocm_fa_library()} fully_masked_rows=0..{masked-1}; NaN patterns over {n} runs (count_rows, first, last, lse_nan, lse_inf): {seen if seen else 'none'}")
```

`gpu_cotenant.py`:

```python
import torch, time, sys
# continuous GEMM + memory churn on the current GPU until killed
a = torch.randn(6144, 6144, device="cuda"); b = torch.randn(6144, 6144, device="cuda")
hold = torch.empty(2 << 30, dtype=torch.uint8, device="cuda")   # 2 GiB resident
while True:
    for _ in range(20): c = a @ b
    x = torch.empty(256 << 20, dtype=torch.uint8, device="cuda"); del x   # allocator churn
    torch.cuda.synchronize()
```

## Suggested fix

Move `continue_condition = True` inside the `while` loop (set it at the top of every iteration), or replace the flag with per-iteration control flow, so that an early exit affects only the tile that triggered it. Backward (`bwd_kernel_fuse.py`) should be checked for the same pattern if it shares the persistent-loop structure.

## Impact in PyTorch

Any memory-efficient (AOTriton) attention with a bottom-right causal mask and `seqlen_q > seqlen_k` whose batch x heads x query tiles exceeds the workgroup count silently returns unwritten output rows; the PyTorch test that catches the small-shape variant is `test/test_transformers.py::TestSDPAAccelerator::test_mem_efficient_bottom_right_fully_masked_rows` on ROCm (kept skipped with this reason).

This report was prepared with the assistance of an AI agent (Claude); the analysis, reproducers and numbers were verified by hand.
